internal/worker_versioning_info.go

```diff
@@ -0,0 +1,310 @@
+		// (task reachability is not reported per task type).
+		ReportTaskReachability bool
+	}
+
+	// WorkerVersionCapabilities includes a worker's build identifier
+	// and whether it is choosing to use the versioning feature.
+	// It is an optional component of [PollerInfo].
+	WorkerVersionCapabilities struct {
+		BuildID       string
+		UseVersioning bool
+	}
+
+	// PollerInfo provides information about a worker/client polling a task queue.
+	// It is used by [TaskQueueTypeInfo].
+	PollerInfo struct {
+		// Time of the last poll. A value of zero means it was not set.
+		LastAccessTime time.Time
+		// The identity of the worker/client who is polling this task queue.
+		Identity string
+		// Polling rate. A value of zero means it was not set.
+		RatePerSecond float64
+		// Optional poller versioning capabilities. Available when a worker has opted into the worker versioning feature.
+		WorkerVersionCapabilities *WorkerVersionCapabilities
+	}
+
+	// TaskQueueTypeInfo specifies task queue information per task type and Build ID.
+	// It is included in [TaskQueueVersionInfo].
+	TaskQueueTypeInfo struct {
+		// Poller details for this task queue category.
+		Pollers []PollerInfo
+	}
+
+	// TaskQueueVersionInfo includes task queue information per Build ID.
+	// It is part of [TaskQueueInfo].
+	TaskQueueVersionInfo struct {
+		// Task queue info per task type.
+		TypesInfo map[TaskQueueType]TaskQueueTypeInfo
+		// The category of tasks that may reach a versioned worker of a certain Build ID.
+		TaskReachability BuildIDTaskReachability
+	}
+
+	// TaskQueueInfo is the response to [Client.DescribeTaskQueueEnhanced].
+	TaskQueueInfo struct {
+		// Task queue information for each Build ID. Empty string as key value means unversioned.
+		VersionsInfo map[string]TaskQueueVersionInfo
+	}
+)
+
+func (o *DescribeTaskQueueEnhancedOptions) validateAndConvertToProto(namespace string) (*workflowservice.DescribeTaskQueueRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+
+	if o.TaskQueue == "" {
+		return nil, errors.New("missing task queue field")
+	}
+
+	taskQueueTypes := make([]enumspb.TaskQueueType, len(o.TaskQueueTypes))
+	for i, t := range o.TaskQueueTypes {
+		taskQueueTypes[i] = taskQueueTypeToProto(t)
+	}
+
+	opt := &workflowservice.DescribeTaskQueueRequest{
+		Namespace: namespace,
+		TaskQueue: &taskqueuepb.TaskQueue{
+			// Sticky queues not supported
+			Name: o.TaskQueue,
+		},
+		ApiMode:                enumspb.DESCRIBE_TASK_QUEUE_MODE_ENHANCED,
+		Versions:               taskQueueVersionSelectionToProto(o.Versions),
+		TaskQueueTypes:         taskQueueTypes,
+		ReportPollers:          o.ReportPollers,
+		ReportTaskReachability: o.ReportTaskReachability,
+	}
+
+	return opt, nil
+}
+
+func workerVersionCapabilitiesFromResponse(response *common.WorkerVersionCapabilities) *WorkerVersionCapabilities {
+	if response == nil {
+		return nil
+	}
```

Comment:
```suggestion
```

Not needed, those get calls below are nil-safe

Reply:
Yes, but I think we want to return a nil pointer as opposed to an empty `&WorkerVersionCapabilities{}`

Reply:
Ah, true! My mistake.